### PR TITLE
Linear scheduler

### PR DESCRIPTION
**Pull request by amunozo**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -131,8 +131,10 @@ def train(args, lm, dataset, tp, dp, saving_folder):
         # device, else it runs all of main.py n_devices times (????).
         # presumably its for multi-gpu training but i haven't learned how yet
         max_epochs=tp.epochs, val_check_interval=tp.val_check_epoch_frac)
-
-    mytrainer = Trainer(lm, tp, start_time=start_time)
+    
+    mytrainer = Trainer(lm, tp, 
+                        expected_batches_per_epoch=len(dataset.train_dataloader(tp.batch_size)),
```
Comment (gailweiss) [inline main.py:136]: this code assumes dataset.train_dataloader will return a dataloader of the same length every time you call it - in particular, that the length here will be the same length as the actual dataloader sent for training (generated below, line 240). that happens to be true but i dont like assuming.
to remove the assumption, create the train dataloader once (e.g. in line 134, with `tdl = dataset.train_dataloader(tp.batch_size)`), and then use the same one each time (i.e. in the initiation of mytrainer, and in the call to pltrainer.fit ).

File: train/train_params.py [hunk 2]
```diff
@@ -109,4 +113,8 @@ def make_tp(forgiving=False, takes_extras=False, convert_lists_to_tuples=False,
 #   Relevant only when also passing --save through the args, otherwise will be
 #   overridden to 0.
 # early_stop_nsamples:
-#   When >0, the number of batches to train before early stopping the training.
\ No newline at end of file
+#   When >0, the number of batches to train before early stopping the training.
+# weight_decay:
+#   Weight decay to apply to the optimizer.
```
Comment (gailweiss) [inline train/train_params.py:118]: stray - weight decay and random seed shouldnt be in here

File: train/trainer.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 class Trainer(pl.LightningModule):
     def __init__(self, model, train_params, start_time=None,
-                 samples_at_validation=True):
+                 samples_at_validation=True, expected_batches_per_epoch=None):
```
Comment (gailweiss) [inline train/trainer.py:10]: if goes past 80 characters, put on new line

Comment (gailweiss) [reply train/trainer.py:10]: (and anywhere else that goes over 80 characters)

File: train/trainer.py [hunk 4]
```diff
@@ -231,25 +232,31 @@ def make_main_scheduler(self, optimizer):
         if self.train_params.lr_scheduler_type == 'Plateau':
             sched = torch.optim.lr_scheduler.ReduceLROnPlateau
             return sched(optimizer, mode='min', verbose=False,
-                         min_lr=self.train_params.min_lr,
-                         factor=self.train_params.scheduler_factor,
-                         patience=self.train_params.patience)
+                        min_lr=self.train_params.min_lr,
```
Comment (gailweiss) [inline train/trainer.py:235]: did a pep8 style checker suggest this, or is this a leftover from earlier?

File: train/trainer.py [hunk 5]
```diff
@@ -231,25 +232,31 @@ def make_main_scheduler(self, optimizer):
...
+                        factor=self.train_params.scheduler_factor,
+                        patience=self.train_params.patience)
         elif self.train_params.lr_scheduler_type == 'Cyclic':
             sched = torch.optim.lr_scheduler.CyclicLR
             return sched(optimizer, base_lr=self.train_params.min_lr,
-                         max_lr=self.train_params.lr,
-                         step_size_up=self.train_params.lr_cycle_steps // 2,
-                         mode='triangular',
-                         gamma=self.train_params.scheduler_factor,
-                         cycle_momentum=False)
```
Comment (gailweiss) [inline train/trainer.py:244]: same q - is this from a style checker or is it a leftover?

File: train/trainer.py [hunk 6]
```diff
@@ -231,25 +232,31 @@ def make_main_scheduler(self, optimizer):
...
         elif self.train_params.lr_scheduler_type == 'Cosine':
             sched = torch.optim.lr_scheduler.CosineAnnealingLR
             return sched(optimizer, self.train_params.lr_cycle_steps,
-                         eta_min=self.train_params.min_lr)
+                        eta_min=self.train_params.min_lr)
+        elif self.train_params.lr_scheduler_type == 'Linear':
+            sched = torch.optim.lr_scheduler.LinearLR
+            return sched(optimizer, start_factor=1.0,
+                        end_factor=self.train_params.min_lr / self.train_params.lr,
+                        total_iters=total_iters)
```
Comment (gailweiss) [inline train/trainer.py:254]: definition of total_iters not present in this commit! also, i would call it total_expected_steps

File: train/trainer.py [hunk 7]
```diff
@@ -231,25 +232,31 @@ def make_main_scheduler(self, optimizer):
...
         else:
             raise Exception("unknown scheduler type:",
                             self.train_params.lr_scheduler_type)
 
+
```
Comment (gailweiss) [inline train/trainer.py:259]: remove added line - pep8 dictates: 2 lines between functions/classes only for top-level functions/classes in the file!

File: train/trainer.py (hunk 4 above)
Comment (amunozo) [reply train/trainer.py:235]: leftover

File: train/trainer.py [hunk 8]
```diff
@@ -250,9 +250,15 @@ def make_main_scheduler(self, optimizer):
                          eta_min=self.train_params.min_lr)
         elif self.train_params.lr_scheduler_type == 'Linear':
             sched = torch.optim.lr_scheduler.LinearLR
+            total_expected_batches = (
+                self.train_params.epochs *
+                self.expected_batches_per_epoch *
+                self.train_params.accumulate_grad_batches -
```
Comment (gailweiss) [inline train/trainer.py:256]: accumulate_grad_batches joins multiple batches to act as one. e.g. if you have accumulate_grad_batches=2, and your dataloader creates 6 batches [a b c d e f] in that order, you'll effectively have 3 batches (a,b), (c,d), (e,f). so it actually reduces not increases the number of effective batches you have. also given we have a disparity between dataloader batches and "stepped" batches, maybe use clearer names than "batches" so there's no confusion: e.g. `train_dataloader_nbatches` instead of  `expected_batches_per_epoch`, and `expected_main_scheduler_steps` for what you're computing here. `expected_main_scheduler_steps = ((self.train_params.epochs * self.train_dataloader_nbatches) // self.train_params.accumulate_grad_batches) - self.train_params.lr_warm_steps` 

this could be smaller than the true number by one if it doesn't divide perfectly by accumulate_grad_batches, but given this is just for scheduling the learning rate, i think thats fine

File: train/train_params.py [hunk 9]
```diff
@@ -109,4 +113,6 @@ def make_tp(forgiving=False, takes_extras=False, convert_lists_to_tuples=False,
 #   Relevant only when also passing --save through the args, otherwise will be
 #   overridden to 0.
 # early_stop_nsamples:
-#   When >0, the number of batches to train before early stopping the training.
\ No newline at end of file
+#   When >0, the number of batches to train before early stopping the training.
+# random_seed:
```
Comment (gailweiss) [inline train/train_params.py:117]: random seed addition needs to go
